edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/brokerConnection/MqttBrokerConnector.cs

```diff
@@ -24,13 +24,15 @@ public class MqttBrokerConnector : IMqttBrokerConnector
 
         readonly object guard = new object();
 
+        readonly TaskCompletionSource<bool> onConnectedTcs = new TaskCompletionSource<bool>();
+
         Option<Channel<MqttPublishInfo>> publications;
         Option<Task> forwardingLoop;
         Option<MqttClient> mqttClient;
 
         AtomicBoolean isRetrying = new AtomicBoolean(false);
 
-        public event EventHandler OnConnected;
+        public Task EnsureConnected => this.onConnectedTcs.Task;
```

Comment:
Event triggers when the connector gets connected, but those who 'came late' will not notice it (they subscribe later to the event, so they will not get the event notification). Using task, once the connector connected, the task will be completed, so using an

await EnsureConnected

will pass through if the connection already happened, or blocks (and awaits) when it is not connected yet.

This mechanism was not needed for a long time, now it is needed because the upstream communication is also goes through the connector, and once edgeHub starts, it immediately tries to talk upstream, but the order is unpredictable (that whether the connector connects first or edgeHub starts talking first) This task was added to solve this issue, but then the rest of the code inherited this usage.

The event was added by others where during dependency resolution their classes was created among the first things, so it was not an issue that the connector connects sooner and they miss the event.

Reply:
makes sense